src/ray/core_worker/core_worker.cc

```diff
@@ -832,6 +832,11 @@ void CoreWorker::Exit(
   RAY_CHECK_OK(
       local_raylet_client_->NotifyDirectCallTaskBlocked(/*release_resources*/ true));
 
+  RAY_LOG(DEBUG) << "Exit signal received, remove all local references.";
+  /// Since this core worker is exiting, it's necessary to release all local references,
+  /// otherwise the frontend code may not release its references and this worker will be
+  /// leaked. See https://github.com/ray-project/ray/issues/19639.
+  reference_counter_->ReleaseAllLocalReferences();
```

Comment:
Sorry I don't have much context here, but if we do this always does `reference_counter_->DrainAndShutdown(shutdown);` make sense?

Reply:
Is this method guaranteed to be called only upon SIGTERM?

Reply:
> Is this method guaranteed to be called only upon SIGTERM?

It will be called in cases of SIGTERM, receiving task_canceled, or other things which invoke `CoreWorker::Exit`

> Sorry I don't have much context here, but if we do this always does reference_counter_-> DrainAndShutdown(shutdown); make sense?

I can't make sure whether `drain_references_callback` is able to be called since it's in another callback.
In current approach, when worker is exiting, it means the local references should be released always. (For task exits normally, the local refs were released in the frontend.)

Reply:
It is called right? We don't shutdown until actual `shutdown` callback is invoked.

cc @stephanie-wang would this be safe for features like "not killing a worker when we own references"?

Reply:
Maybe we should do this only for actor cases?

Reply:
I believe this is correct. You can check out the test that @jovany-wang added for an example where this breaks even though no actors are used.

This will still keep the object in scope if there are other workers using the ref. It's just to make sure that we don't consider the owner's own refs anymore (which can't get released otherwise because we won't return to Python).

Reply:
I see. To make sure I understood this code

```
    for (int i = ref.second.local_ref_count; i > 0; --i) {
      RemoveLocalReferenceInternal(ref.first, nullptr);
    }
```

Only removes the owner's ref count, so if it is used by other workers, it won't be stopped right? (And the purpose of DrainAndShutdown is for this case, so it won't regress)